lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Rules.hs

```diff
@@ -5,16 +5,22 @@ module LambdaBuffers.Compiler.TypeClass.Rules (
   Class (..),
   Constraint (..),
   Rule (..),
-  type Instance,
+  FQClassName (..),
   mapPat,
+  ruleHeadPat,
+  ruleHeadClass,
 ) where
 
-import LambdaBuffers.Compiler.ProtoCompat qualified as P
+import Data.Text (Text)
+
 import LambdaBuffers.Compiler.TypeClass.Pat (Pat)
 
+data FQClassName = FQClassName {cName :: Text, cModule :: [Text]}
+  deriving stock (Show, Eq, Ord)
```

Comment:
I think the reason you need this is because your `solve` doesn't have access to any 'context'. That's ok, but introduces separate processing steps which is not really necessary.

TyClassRef would be sufficient if your solver worked with
 ```haskell
 SolveContext :: '[ "currentModule" := ModuleName, "classDefs" :=  Map TyClassName TyClassDef]
 ```

Reply:
`solve` isn't supposed to have access to a context. It's supposed to be the simplest possible implementation of the constraint solver algorithm, which is used as the central component later on when a context (or something that's effectively the same) gets constructed. 

There's no hard theoretical reason for that - `solve` was just hard to debug so I cut it down to the bare essentials. 

In my mind, `FQClassName` was just a safety valve to make sure we don't waste time creating/working with classes that differ only in the `SourceInfo` of their `TyClassRef`. I guess I could use `TyClassRef` and a slightly degenerate Eq instance, /shrug

Reply:
1. No need to do the Eq
2. I told you SourceInfo should have been a type parameter :D
3. 'solve' does have access to a context, the first argument is a ruleset it can observe right? That's the context, I'm just saying it doesn't have to be the only thing in the context.